File: integration/bunjs-only-snippets/ffi.test.js
import { describe, it, expect } from "bun:test";
import { unsafe } from "bun";
//
import {
  native,
  viewSource,
  dlopen,
  CString,
  ptr,
  toBuffer,
  toArrayBuffer,
  FFIType,
  callback,
} from "bun:ffi";

it("ffi print", async () => {
  await Bun.write(
    import.meta.dir + "/ffi.test.fixture.callback.c",
    viewSource(
      {
        return_type: "bool",
        args: ["ptr"],
      },
      true
    )
  );
  await Bun.write(
    import.meta.dir + "/ffi.test.fixture.receiver.c",
    viewSource(
      {
        not_a_callback: {
          return_type: "float",
          args: ["float"],
        },
      },
      false
    )[0]
  );
  expect(
    viewSource(
      {
        return_type: "int8_t",
        args: [],
      },
      true
    ).length > 0
  ).toBe(true);
  expect(
    viewSource(
      {
        a: {
          return_type: "int8_t",
          args: [],
        },
      },
      false
    ).length > 0
  ).toBe(true);
});

it("ffi run", () => {
  const types = {
    returns_true: {
      return_type: "bool",
      args: [],
    },
    returns_false: {
      return_type: "bool",
      args: [],
    },
    returns_42_char: {
      return_type: "char",
      args: [],
    },
    returns_42_float: {
      return_type: "float",
      args: [],
    },
    returns_42_double: {
      return_type: "double",
      args: [],
    },
    returns_42_uint8_t: {
      return_type: "uint8_t",
      args: [],
    },
    returns_neg_42_int8_t: {
      return_type: "int8_t",
      args: [],
    },
    returns_42_uint16_t: {
      return_type: "uint16_t",
      args: [],
    },
    returns_42_uint32_t: {
      return_type: "uint32_t",
      args: [],
    },
    returns_42_uint64_t: {
      return_type: "uint64_t",
      args: [],
    },
    returns_neg_42_int16_t: {
      return_type: "int16_t",
      args: [],
    },
    returns_neg_42_int32_t: {
      return_type: "int32_t",
      args: [],
    },
    returns_neg_42_int64_t: {
      return_type: "int64_t",
      args: [],
    },

    identity_char: {
      return_type: "char",
      args: ["char"],
    },
    identity_float: {
      return_type: "float",
      args: ["float"],
    },
    identity_bool: {
      return_type: "bool",
      args: ["bool"],
    },
    identity_double: {
      return_type: "double",
      args: ["double"],
    },
    identity_int8_t: {
      return_type: "int8_t",
      args: ["int8_t"],
    },
    identity_int16_t: {
      return_type: "int16_t",
      args: ["int16_t"],
    },
    identity_int32_t: {
      return_type: "int32_t",
      args: ["int32_t"],
    },
    identity_int64_t: {
      return_type: "int64_t",
      args: ["int64_t"],
    },
    identity_uint8_t: {
      return_type: "uint8_t",
      args: ["uint8_t"],
    },
    identity_uint16_t: {
      return_type: "uint16_t",
      args: ["uint16_t"],
    },
    identity_uint32_t: {
      return_type: "uint32_t",
      args: ["uint32_t"],
    },
    identity_uint64_t: {
      return_type: "uint64_t",
      args: ["uint64_t"],
    },

    add_char: {
      return_type: "char",
      args: ["char", "char"],
    },
    add_float: {
      return_type: "float",
      args: ["float", "float"],
    },
    add_double: {
      return_type: "double",
      args: ["double", "double"],
    },
    add_int8_t: {
      return_type: "int8_t",
      args: ["int8_t", "int8_t"],
    },
    add_int16_t: {
      return_type: "int16_t",
      args: ["int16_t", "int16_t"],
    },
    add_int32_t: {
      return_type: "int32_t",
      args: ["int32_t", "int32_t"],
    },
    add_int64_t: {
      return_type: "int64_t",
      args: ["int64_t", "int64_t"],
    },
    add_uint8_t: {
      return_type: "uint8_t",
      args: ["uint8_t", "uint8_t"],
    },
    add_uint16_t: {
      return_type: "uint16_t",
      args: ["uint16_t", "uint16_t"],
    },
    add_uint32_t: {
      return_type: "uint32_t",
      args: ["uint32_t", "uint32_t"],
    },

    does_pointer_equal_42_as_int32_t: {
      return_type: "bool",
      args: ["ptr"],
    },

    ptr_should_point_to_42_as_int32_t: {
      return_type: "ptr",
      args: [],
    },
    identity_ptr: {
      return_type: "ptr",
      args: ["ptr"],
    },
    add_uint64_t: {
      return_type: "uint64_t",
      args: ["uint64_t", "uint64_t"],
    },

    cb_identity_true: {
      return_type: "bool",
      args: ["ptr"],
    },
    cb_identity_false: {
      return_type: "bool",
      args: ["ptr"],
    },
    cb_identity_42_char: {
      return_type: "char",
      args: ["ptr"],
    },
    cb_identity_42_float: {
      return_type: "float",
      args: ["ptr"],
    },
    cb_identity_42_double: {
      return_type: "double",
      args: ["ptr"],
    },
    cb_identity_42_uint8_t: {
      return_type: "uint8_t",
      args: ["ptr"],
    },
    cb_identity_neg_42_int8_t: {
      return_type: "int8_t",
      args: ["ptr"],
    },
    cb_identity_42_uint16_t: {
      return_type: "uint16_t",
      args: ["ptr"],
    },
    cb_identity_42_uint32_t: {
      return_type: "uint32_t",
      args: ["ptr"],
    },
    cb_identity_42_uint64_t: {
      return_type: "uint64_t",
      args: ["ptr"],
    },
    cb_identity_neg_42_int16_t: {
      return_type: "int16_t",
      args: ["ptr"],
    },
    cb_identity_neg_42_int32_t: {
      return_type: "int32_t",
      args: ["ptr"],
    },
    cb_identity_neg_42_int64_t: {
      return_type: "int64_t",
      args: ["ptr"],
    },

    return_a_function_ptr_to_function_that_returns_true: {
      return_type: "ptr",
      args: [],
    },
  };
  const {
    symbols: {
      returns_true,
      returns_false,
      return_a_function_ptr_to_function_that_returns_true,
      returns_42_char,
      returns_42_float,
      returns_42_double,
      returns_42_uint8_t,
      returns_neg_42_int8_t,
      returns_42_uint16_t,
      returns_42_uint32_t,
      returns_42_uint64_t,
      returns_neg_42_int16_t,
      returns_neg_42_int32_t,
      returns_neg_42_int64_t,
      identity_char,
      identity_float,
      identity_bool,
      identity_double,
      identity_int8_t,
      identity_int16_t,
      identity_int32_t,
      identity_int64_t,
      identity_uint8_t,
      identity_uint16_t,
      identity_uint32_t,
      identity_uint64_t,
      add_char,
      add_float,
      add_double,
      add_int8_t,
      add_int16_t,
      add_int32_t,
      add_int64_t,
      add_uint8_t,
      add_uint16_t,
      identity_ptr,
      add_uint32_t,
      add_uint64_t,
      does_pointer_equal_42_as_int32_t,
      ptr_should_point_to_42_as_int32_t,
      cb_identity_true,
      cb_identity_false,
      cb_identity_42_char,
      cb_identity_42_float,
      cb_identity_42_double,
      cb_identity_42_uint8_t,
      cb_identity_neg_42_int8_t,
      cb_identity_42_uint16_t,
      cb_identity_42_uint32_t,
      cb_identity_42_uint64_t,
      cb_identity_neg_42_int16_t,
      cb_identity_neg_42_int32_t,
      cb_identity_neg_42_int64_t,
    },
    close,
  } = dlopen("/tmp/bun-ffi-test.dylib", types);

  expect(returns_true()).toBe(true);

  expect(returns_false()).toBe(false);

  expect(returns_42_char()).toBe(42);
  expect(returns_42_uint64_t().valueOf()).toBe(42);

  expect(Math.fround(returns_42_float())).toBe(Math.fround(42.41999804973602));
  expect(returns_42_double()).toBe(42.42);
  expect(returns_42_uint8_t()).toBe(42);
  expect(returns_neg_42_int8_t()).toBe(-42);
  expect(returns_42_uint16_t()).toBe(42);
  expect(returns_42_uint32_t()).toBe(42);
  expect(returns_42_uint64_t()).toBe(42);
  expect(returns_neg_42_int16_t()).toBe(-42);
  expect(returns_neg_42_int32_t()).toBe(-42);
  expect(identity_int32_t(10)).toBe(10);
  expect(returns_neg_42_int64_t()).toBe(-42);

  expect(identity_char(10)).toBe(10);

  expect(identity_float(10.199999809265137)).toBe(10.199999809265137);

  expect(identity_bool(true)).toBe(true);

  expect(identity_bool(false)).toBe(false);
  expect(identity_double(10.100000000000364)).toBe(10.100000000000364);

  expect(identity_int8_t(10)).toBe(10);
  expect(identity_int16_t(10)).toBe(10);
  expect(identity_int64_t(10)).toBe(10);
  expect(identity_uint8_t(10)).toBe(10);
  expect(identity_uint16_t(10)).toBe(10);
  expect(identity_uint32_t(10)).toBe(10);
  expect(identity_uint64_t(10)).toBe(10);

  var bigArray = new BigUint64Array(8);
  new Uint8Array(bigArray.buffer).fill(255);
  var bigIntArray = new BigInt64Array(bigArray.buffer);

  expect(identity_uint64_t(bigArray[0])).toBe(bigArray[0]);
  expect(identity_uint64_t(bigArray[0] - BigInt(1))).toBe(
    bigArray[0] - BigInt(1)
  );
  expect(add_uint64_t(BigInt(-1) * bigArray[0], bigArray[0])).toBe(0);
  expect(add_uint64_t(BigInt(-1) * bigArray[0] + BigInt(10), bigArray[0])).toBe(
    10
  );
  expect(identity_uint64_t(0)).toBe(0);
  expect(identity_uint64_t(100)).toBe(100);
  expect(identity_uint64_t(BigInt(100))).toBe(100);
  expect(identity_int64_t(bigIntArray[0])).toBe(bigIntArray[0]);
  expect(identity_int64_t(bigIntArray[0] - BigInt(1))).toBe(
    bigIntArray[0] - BigInt(1)
  );

  expect(add_char(1, 1)).toBe(2);
  expect(add_float(2.4, 2.8)).toBe(Math.fround(5.2));
  expect(add_double(4.2, 0.1)).toBe(4.3);
  expect(add_int8_t(1, 1)).toBe(2);
  expect(add_int16_t(1, 1)).toBe(2);
  expect(add_int32_t(1, 1)).toBe(2);
  expect(add_int64_t(1, 1)).toBe(2);
  expect(add_uint8_t(1, 1)).toBe(2);
  expect(add_uint16_t(1, 1)).toBe(2);
  expect(add_uint32_t(1, 1)).toBe(2);

  const cptr = ptr_should_point_to_42_as_int32_t();
  expect(cptr != 0).toBe(true);
  expect(typeof cptr === "number").toBe(true);
  expect(does_pointer_equal_42_as_int32_t(cptr)).toBe(true);
  const buffer = toBuffer(cptr, 0, 4);
  expect(buffer.readInt32(0)).toBe(42);
  expect(new DataView(toArrayBuffer(cptr, 0, 4), 0, 4).getInt32(0, true)).toBe(
    42
  );
  expect(ptr(buffer)).toBe(cptr);
  expect(new CString(cptr, 0, 1).toString()).toBe("*");
  expect(identity_ptr(cptr)).toBe(cptr);
  const second_ptr = ptr(new Buffer(8));
  expect(identity_ptr(second_ptr)).toBe(second_ptr);
  // function identityBool() {
  //   return true;
  // }
  // globalThis.identityBool = identityBool;

  // const first = native.callback(
  //   {
  //     return_type: "bool",
  //   },
  //   identityBool
  // );
  // expect(
  //   cb_identity_true(return_a_function_ptr_to_function_that_returns_true())
  // ).toBe(true);

  // expect(cb_identity_true(first)).toBe(true);

  // expect(
  //   cb_identity_false(
  //     callback(
  //       {
  //         return_type: "bool",
  //       },
  //       () => false
  //     )
  //   )
  // ).toBe(false);

  // expect(
  //   cb_identity_42_char(
  //     callback(
  //       {
  //         return_type: "char",
  //       },
  //       () => 42
  //     )
  //   )
  // ).toBe(42);
  // expect(
  //   cb_identity_42_uint8_t(
  //     callback(
  //       {
  //         return_type: "uint8_t",
  //       },
  //       () => 42
  //     )
  //   )
  // ).toBe(42);

  // cb_identity_neg_42_int8_t(
  //   callback(
  //     {
  //       return_type: "int8_t",
  //     },
  //     () => -42
  //   )
  // ).toBe(-42);

  // cb_identity_42_uint16_t(
  //   callback(
  //     {
  //       return_type: "uint16_t",
  //     },
  //     () => 42
  //   )
  // ).toBe(42);

  // cb_identity_42_uint32_t(
  //   callback(
  //     {
  //       return_type: "uint32_t",
  //     },
  //     () => 42
  //   )
  // ).toBe(42);

  // cb_identity_neg_42_int16_t(
  //   callback(
  //     {
  //       return_type: "int16_t",
  //     },
  //     () => -42
  //   )
  // ).toBe(-42);

  // cb_identity_neg_42_int32_t(
  //   callback(
  //     {
  //       return_type: "int32_t",
  //     },
  //     () => -42
  //   )
  // ).toBe(-42);

  close();
});

File: test/js/bun/ffi/ffi-test.c
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#ifdef _WIN32
#define FFI_EXPORT __declspec(dllexport)
#else
#define FFI_EXPORT __attribute__((visibility("default")))
#endif

FFI_EXPORT bool returns_true();
FFI_EXPORT bool returns_false();
FFI_EXPORT char returns_42_char();
FFI_EXPORT float returns_42_float();
FFI_EXPORT double returns_42_double();
FFI_EXPORT uint8_t returns_42_uint8_t();
FFI_EXPORT int8_t returns_neg_42_int8_t();
FFI_EXPORT uint16_t returns_42_uint16_t();
FFI_EXPORT uint32_t returns_42_uint32_t();
FFI_EXPORT uint64_t returns_42_uint64_t();
FFI_EXPORT int16_t returns_neg_42_int16_t();
FFI_EXPORT int32_t returns_neg_42_int32_t();
FFI_EXPORT int64_t returns_neg_42_int64_t();

FFI_EXPORT bool cb_identity_true(bool (*cb)());
FFI_EXPORT bool cb_identity_false(bool (*cb)());
FFI_EXPORT char cb_identity_42_char(char (*cb)());
FFI_EXPORT float cb_identity_42_float(float (*cb)());
FFI_EXPORT double cb_identity_42_double(double (*cb)());
FFI_EXPORT uint8_t cb_identity_42_uint8_t(uint8_t (*cb)());
FFI_EXPORT int8_t cb_identity_neg_42_int8_t(int8_t (*cb)());
FFI_EXPORT uint16_t cb_identity_42_uint16_t(uint16_t (*cb)());
FFI_EXPORT uint32_t cb_identity_42_uint32_t(uint32_t (*cb)());
FFI_EXPORT uint64_t cb_identity_42_uint64_t(uint64_t (*cb)());
FFI_EXPORT int16_t cb_identity_neg_42_int16_t(int16_t (*cb)());
FFI_EXPORT int32_t cb_identity_neg_42_int32_t(int32_t (*cb)());
FFI_EXPORT int64_t cb_identity_neg_42_int64_t(int64_t (*cb)());

FFI_EXPORT bool identity_bool_true();
FFI_EXPORT bool identity_bool_false();
FFI_EXPORT char identity_char(char a);
FFI_EXPORT float identity_float(float a);
FFI_EXPORT bool identity_bool(bool ident);
FFI_EXPORT double identity_double(double a);
FFI_EXPORT int8_t identity_int8_t(int8_t a);
FFI_EXPORT int16_t identity_int16_t(int16_t a);
FFI_EXPORT int32_t identity_int32_t(int32_t a);
FFI_EXPORT int64_t identity_int64_t(int64_t a);
FFI_EXPORT uint8_t identity_uint8_t(uint8_t a);
FFI_EXPORT uint16_t identity_uint16_t(uint16_t a);
FFI_EXPORT uint32_t identity_uint32_t(uint32_t a);
FFI_EXPORT uint64_t identity_uint64_t(uint64_t a);
FFI_EXPORT void *identity_ptr(void *ident);

FFI_EXPORT char add_char(char a, char b);
FFI_EXPORT float add_float(float a, float b);
FFI_EXPORT double add_double(double a, double b);
FFI_EXPORT int8_t add_int8_t(int8_t a, int8_t b);
FFI_EXPORT int16_t add_int16_t(int16_t a, int16_t b);
FFI_EXPORT int32_t add_int32_t(int32_t a, int32_t b);
FFI_EXPORT int64_t add_int64_t(int64_t a, int64_t b);
FFI_EXPORT uint8_t add_uint8_t(uint8_t a, uint8_t b);
FFI_EXPORT uint16_t add_uint16_t(uint16_t a, uint16_t b);
FFI_EXPORT uint32_t add_uint32_t(uint32_t a, uint32_t b);
FFI_EXPORT uint64_t add_uint64_t(uint64_t a, uint64_t b);

bool returns_false() { return false; }
bool returns_true() { return true; }
char returns_42_char() { return '*'; }
double returns_42_double() { return (double)42.42; }
float returns_42_float() { return 42.42f; }
int16_t returns_neg_42_int16_t() { return -42; }
int32_t returns_neg_42_int32_t() { return -42; }
int64_t returns_neg_42_int64_t() { return -42; }
int8_t returns_neg_42_int8_t() { return -42; }
uint16_t returns_42_uint16_t() { return 42; }
uint32_t returns_42_uint32_t() { return 42; }
uint64_t returns_42_uint64_t() { return 42; }
uint8_t returns_42_uint8_t() { return (uint8_t)42; }

char identity_char(char a) { return a; }
float identity_float(float a) { return a; }
double identity_double(double a) { return a; }
int8_t identity_int8_t(int8_t a) { return a; }
int16_t identity_int16_t(int16_t a) { return a; }
int32_t identity_int32_t(int32_t a) { return a; }
int64_t identity_int64_t(int64_t a) { return a; }
uint8_t identity_uint8_t(uint8_t a) { return a; }
uint16_t identity_uint16_t(uint16_t a) { return a; }
uint32_t identity_uint32_t(uint32_t a) { return a; }
uint64_t identity_uint64_t(uint64_t a) { return a; }
bool identity_bool(bool ident) { return ident; }
void *identity_ptr(void *ident) { return ident; }

char add_char(char a, char b) { return a + b; }
float add_float(float a, float b) { return a + b; }
double add_double(double a, double b) { return a + b; }
int8_t add_int8_t(int8_t a, int8_t b) { return a + b; }
int16_t add_int16_t(int16_t a, int16_t b) { return a + b; }
int32_t add_int32_t(int32_t a, int32_t b) { return a + b; }
int64_t add_int64_t(int64_t a, int64_t b) { return a + b; }
uint8_t add_uint8_t(uint8_t a, uint8_t b) { return a + b; }
uint16_t add_uint16_t(uint16_t a, uint16_t b) { return a + b; }
uint32_t add_uint32_t(uint32_t a, uint32_t b) { return a + b; }
uint64_t add_uint64_t(uint64_t a, uint64_t b) { return a + b; }

FFI_EXPORT void *ptr_should_point_to_42_as_int32_t();

void *ptr_should_point_to_42_as_int32_t() {
  int32_t *ptr = malloc(sizeof(int32_t));
  *ptr = 42;
  return ptr;
}

static uint8_t buffer_with_deallocator[128];
static int deallocatorCalled;
FFI_EXPORT void deallocator(void *ptr, void *userData) { deallocatorCalled++; }
FFI_EXPORT void *getDeallocatorCallback() {
  deallocatorCalled = 0;
  return &deallocator;
}
FFI_EXPORT void *getDeallocatorBuffer() {
  deallocatorCalled = 0;
  return &buffer_with_deallocator;
}
FFI_EXPORT int getDeallocatorCalledCount() { return deallocatorCalled; }

FFI_EXPORT bool is_null(int32_t *ptr) { return ptr == NULL; }
FFI_EXPORT bool does_pointer_equal_42_as_int32_t(int32_t *ptr);
bool does_pointer_equal_42_as_int32_t(int32_t *ptr) { return *ptr == 42; }

FFI_EXPORT void *return_a_function_ptr_to_function_that_returns_true();
void *return_a_function_ptr_to_function_that_returns_true() {
  return (void *)&returns_true;
}

FFI_EXPORT bool cb_identity_true(bool (*cb)()) { return cb(); }

FFI_EXPORT bool cb_identity_false(bool (*cb)()) { return cb(); }
FFI_EXPORT char cb_identity_42_char(char (*cb)()) { return cb(); }
FFI_EXPORT float cb_identity_42_float(float (*cb)()) { return cb(); }
FFI_EXPORT double cb_identity_42_double(double (*cb)()) { return cb(); }
FFI_EXPORT uint8_t cb_identity_42_uint8_t(uint8_t (*cb)()) { return cb(); }
FFI_EXPORT int8_t cb_identity_neg_42_int8_t(int8_t (*cb)()) { return cb(); }
FFI_EXPORT uint16_t cb_identity_42_uint16_t(uint16_t (*cb)()) { return cb(); }
FFI_EXPORT uint32_t cb_identity_42_uint32_t(uint32_t (*cb)()) { return cb(); }
FFI_EXPORT uint64_t cb_identity_42_uint64_t(uint64_t (*cb)()) { return cb(); }
FFI_EXPORT int16_t cb_identity_neg_42_int16_t(int16_t (*cb)()) { return cb(); }
FFI_EXPORT int32_t cb_identity_neg_42_int32_t(int32_t (*cb)()) { return cb(); }
FFI_EXPORT int64_t cb_identity_neg_42_int64_t(int64_t (*cb)()) { return cb(); }